<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [6]:
import pandas as pd
import psycopg2

In [7]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 


In [8]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [9]:
# текст запроса
query_3_1 = f'''
                SELECT count(*)
                FROM public.vacancies
'''

In [10]:
# результат запроса

result_3_1 = pd.read_sql_query(query_3_1, connection)
print(result_3_1.iloc[0, 0])

49197


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\2322379249.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_1 = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [11]:
# текст запроса
query_3_2 = f'''
                SELECT count(*)
                FROM public.employers
'''

In [12]:
# результат запроса
result_3_2 = pd.read_sql_query(query_3_2, connection)
print(result_3_2.iloc[0, 0])

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\805946276.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_2 = pd.read_sql_query(query_3_2, connection)


23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [13]:
# текст запроса
query_3_3 = f'''
                SELECT count(*)
                FROM public.areas
'''

In [14]:
# результат запроса
result_3_3 = pd.read_sql_query(query_3_3, connection)
print(result_3_3.iloc[0, 0])

1362


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\945151968.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_3 = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [15]:
# текст запроса
query_3_4 = f'''
                SELECT count(*)
                FROM public.industries
'''

In [16]:
# результат запроса
result_3_4 = pd.read_sql_query(query_3_4, connection)
print(result_3_4.iloc[0, 0])

294


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\675272324.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_4 = pd.read_sql_query(query_3_4, connection)


***

In [17]:
# выводы по предварительному анализу данных

В базе данных количество:
* вакансий - 49197
* работодателей - 23501
* регионов - 1362
* сфер деятельности - 294

Количество вакансий больше количества работодателей, следовательно у многих работодателей несколько вакантных мест.

Учитывая что в России всего 1118 городов, а в нашей базе 1362 региона, можно сделать вывод что, вакансии представлены практически во всех городах и регионах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [18]:
# текст запроса
query_4_1 = f'''
                SELECT
                    a.name area,
                    count(*) cnt
                FROM public.vacancies v
                JOIN public.areas a ON v.area_id = a.id --присоединяем таблицу с городами для получения названий
                GROUP BY area                           --группируем по названиям городов для агрегации
                ORDER BY cnt DESC                       --сортировка по количеству вакансий
'''

In [19]:
# результат запроса
result_4_1 = pd.read_sql_query(query_4_1, connection)
print(result_4_1.head())

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\2062383920.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_1 = pd.read_sql_query(query_4_1, connection)


              area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [20]:
# текст запроса
query_4_2 = f'''
                SELECT count(*)
                FROM public.vacancies
                WHERE salary_from IS NOT NULL --условия фильтрации вакансий
                OR salary_to IS NOT NULL      -- -//-
'''

In [21]:
# результат запроса
result_4_2 = pd.read_sql_query(query_4_2, connection)
print(result_4_2.iloc[0, 0])

24073


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1541976964.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_2 = pd.read_sql_query(query_4_2, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [22]:
# текст запроса
query_4_3= f'''
                SELECT
                    AVG(salary_from) lower, --среднее по нижним границам з/п вилки
                    AVG(salary_to) upper    --среднее по верхним границам з/п вилки
                FROM public.vacancies
'''

In [23]:
# результат запроса
result_4_3 = pd.read_sql_query(query_4_3, connection)
print(result_4_3.round())

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1495530319.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_3 = pd.read_sql_query(query_4_3, connection)


     lower     upper
0  71065.0  110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [24]:
# текст запроса
query_4_4= f'''
                SELECT
                    schedule,                 --тип рабочего графика
                    employment,               --тип трудоустройства
                    count(*)                  --количество вакансий
                FROM public.vacancies
                GROUP BY schedule, employment --группируем по типу рабочего графика и трудоустройства, чтобы получить все сочетания
                ORDER BY count(*) DESC        --сортируем по количеству вакансий
'''

In [25]:
# результат запроса
result_4_4 = pd.read_sql_query(query_4_4, connection)
print(result_4_4)

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1531179886.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_4 = pd.read_sql_query(query_4_4, connection)


            schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [26]:
# текст запроса
query_4_5= f'''
                SELECT
                    experience,       --требуемый опыт работы
                    count(*)          --количество вакансий
                FROM public.vacancies
                GROUP BY experience   --группировка по опыту работы
                ORDER BY count(*)
'''

In [27]:
# результат запроса
result_4_5 = pd.read_sql_query(query_4_5, connection)
print(result_4_5)

           experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\3855474113.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_5 = pd.read_sql_query(query_4_5, connection)


***

In [28]:
# выводы по детальному анализу вакансий

В нашей базе 49197 вакансий, большая часть которых распределенна по крупнейшим городам страны, особо выделяется среди них Москва - 5333 вакансии.

Только у половины вакансий заполнено хотябы одно поле зарплатной вилки. Средние значения границ з/п вилки: нижняя - 71065 р., верхняя - 110537 р..

Большинство вакансий требуют: полную занятость на полный день - 35367, на втором месте: полная занятость на удалённую работу - 7802.

В основном компании ищут сотрудников с опытом работы от 1 года до 3 лет: 26152 вакансии, а также более опытных товарищей с опытом работы от 3 до 6 лет: 14511 вакансий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [29]:
# текст запроса
query_5_1 = f'''
                SELECT
                    e.name employer,                            --названия компаний
                    count(*) cnt                                --количество вакансий
                FROM public.vacancies v
                JOIN public.employers e ON v.employer_id = e.id --присоединяем таблицу с работодателями
                GROUP BY employer                               --группировка по названию компании
                ORDER BY cnt DESC
'''

In [30]:
# результат запроса
result_5_1 = pd.read_sql_query(query_5_1, connection)
print(result_5_1.iloc[[0, 4]])

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\499902635.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_1 = pd.read_sql_query(query_5_1, connection)


        employer   cnt
0         Яндекс  1933
4  Газпром нефть   331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [31]:
# текст запроса
query_5_2 = f'''
                SELECT
                    a.name,                                      --название региона
                    count(v.id) vac_cnt,                         --количество вакансий
                    count(e.id) emp_cnt                          --количество работодателей
                FROM public.areas a                              --таблица регионов
                LEFT JOIN public.vacancies v ON v.area_id = a.id --таблицу вакансий подключаем left join, чтобы остались даже те регионы, где нету вакансий
                LEFT JOIN public.employers e ON a.id = e.area    --также подключаем таблицу компаний
                GROUP BY a.id                                    --группировка по регионам
                ORDER BY vac_cnt, emp_cnt DESC
'''

In [32]:
# результат запроса
result_5_2 = pd.read_sql_query(query_5_2, connection)
print(result_5_2.iloc[0, 0])

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\4111559909.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_2 = pd.read_sql_query(query_5_2, connection)


Россия


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [33]:
# текст запроса
query_5_3 = f'''
            SELECT
                name,                                            --название работодателя
                count(*) cnt                                     --количество регионов
            FROM
                (
                 SELECT                                          --делаем подзапрос чтобы вытянуть названия компаний и относящиеся к ним уникальные регионы
                     e.name,                                     --название работодателя
                     v.area_id                                   --id региона
                 FROM public.vacancies v
                 JOIN public.employers e ON e.id = v.employer_id
                 GROUP BY e.id, v.area_id                        --группируем по компании, а также по региону, чтобы остались только уникальные регионы
                ) AS a
            GROUP BY name                                        --группировка по названию компании
            ORDER BY cnt DESC
'''

In [34]:
# результат запроса
result_5_3 = pd.read_sql_query(query_5_3, connection)
print(result_5_3.head())

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\2783423974.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_3 = pd.read_sql_query(query_5_3, connection)


                     name  cnt
0                  Яндекс  181
1              Ростелеком  152
2              Спецремонт  116
3  Поляков Денис Иванович   88
4               ООО ЕФИН    71


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [35]:
# текст запроса
query_5_4 = f'''
                SELECT count(*)
                FROM public.employers e
                LEFT JOIN public.employers_industries ei ON ei.employer_id = e.id   --подключаем таблицу с помощью left join, чтобы остались все компании, даже если сфера деят. не указана
                WHERE industry_id IS NULL                                           --фильтр по сфере деятельности
'''

In [36]:
# результат запроса
result_5_4 = pd.read_sql_query(query_5_4, connection)
print(result_5_4.iloc[0, 0])

8419


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1617712234.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_4 = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [37]:
# текст запроса
query_5_5 = f'''
                SELECT
                    name,
                    count(*)
                FROM public.employers e
                JOIN public.employers_industries ei ON ei.employer_id = e.id
                GROUP BY name                                                   --группировка по названию компании
                HAVING count(*) = 4                                             --фильтр по агрегации
                ORDER BY name
                OFFSET 2                                                        --вывод конкретной строки
                LIMIT 1
'''

In [38]:
# результат запроса
result_5_5 = pd.read_sql_query(query_5_5, connection)
print(result_5_5.iloc[0, 0])

2ГИС


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\651350566.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_5 = pd.read_sql_query(query_5_5, connection)


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [39]:
# текст запроса
query_5_6 = f'''
                SELECT count(*)
                FROM public.employers_industries ei
                JOIN public.industries i ON i.id = ei.industry_id
                WHERE i.name = 'Разработка программного обеспечения' --фильтр по сфере деятельности
'''

In [40]:
# результат запроса
result_5_6 = pd.read_sql_query(query_5_6, connection)
print(result_5_6.iloc[0, 0])

3553


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\2037234874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_6 = pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [41]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)                                                # делаем get запрос
page = BeautifulSoup(response.text, 'html.parser')                          # готовим суп (получаем объект типа BeautifulSoup)

links = page.find('table', 'standard sortable').find('tbody').find_all('a') # поиск по тегам: нужной таблицы -> тело таблицы -> ссылки с названиями городов
city_lst = [link['title'] for link in links[4:]]                            # собираем в список названия городов
print(city_lst)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [42]:
# текст запроса
query_5_7 = f'''
                SELECT
                    a.name,
                    count(v.id) cnt
                FROM public.vacancies v
                JOIN public.areas a ON v.area_id = a.id
                JOIN public.employers e ON v.employer_id = e.id
                WHERE e.name = 'Яндекс' AND a.name IN {tuple(city_lst)} --фильтр по названию компании и городам
                GROUP BY a.name                                         --группировка по городам
                
                UNION ALL
                
                SELECT                                                  --добавляем строку total с общим количеством
                    'Total',
                    count(*)
                FROM public.vacancies v
                JOIN public.areas a ON v.area_id = a.id
                JOIN public.employers e ON v.employer_id = e.id
                WHERE e.name = 'Яндекс' AND a.name IN {tuple(city_lst)}
                
                ORDER BY 2
'''

In [43]:
# результат запроса
result_5_7 = pd.read_sql_query(query_5_7, connection)
print(result_5_7)

               name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1284632481.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_7 = pd.read_sql_query(query_5_7, connection)


***

In [44]:
# выводы по анализу работодателей

In [45]:
# запрос по типу рабочего графика
query_sched = f'''
                SELECT
                    e.name,                                     --названия компаний
                    count(*) cnt                                --количество вакансий
                FROM public.vacancies v
                JOIN public.employers e ON v.employer_id = e.id --присоединяем таблицу с работодателями
                WHERE v.schedule = 'Удаленная работа'           --фильтр по типу рабочего графика
                GROUP BY e.name 
                ORDER BY cnt DESC
'''
result_sched = pd.read_sql_query(query_sched, connection)
print(result_sched.head())

                 name   cnt
0              Яндекс  1690
1             DataArt   165
2            Тинькофф    95
3  Nitka Technologies    77
4           ООО ЕФИН     71


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1113296988.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_sched = pd.read_sql_query(query_sched, connection)


У большинства компаний открыто менее 500 вакансий, очень сильно выделяется среди всех Яндекс - 1933 вакансии. Вероятно это связанно с тем, что  у Яндекса самая широкая география найма, вакансии распределенны по 181 региону, но основная причина в типе рабочего графика, как можно увидеть в запросе выше: у "Яндекса" 1690 вакансий с типом "Удалённая работа", данный список возглавляют в основном IT - компании.

Из 23501 работодателей, у 8419 не указана сфера деятельности, у 3553 работодателей в качестве сферы деятельности указана Разработка программного обеспечения.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [46]:
# текст запроса
query_6_1 = f'''
                SELECT count(*)
                FROM public.vacancies
                WHERE name ILIKE '%data%' --фильтр по названию
                OR name ILIKE '%данн%'
'''

In [47]:
# результат запроса
result_6_1 = pd.read_sql_query(query_6_1, connection)
print(result_6_1.iloc[0, 0])

1771


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\4213675830.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_1 = pd.read_sql_query(query_6_1, connection)


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [48]:
# текст запроса
query_6_2 = f'''
                SELECT count(*)
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']) --фильтр для поиска подходящих по названию вакансий
                OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
                AND (name ILIKE '%junior%' OR experience = 'Нет опыта' OR employment = 'Стажировка')                                                   --дополнительный фильтр для более узкого поиска
'''

In [49]:
# результат запроса
result_6_2 = pd.read_sql_query(query_6_2, connection)
print(result_6_2.iloc[0, 0])

51


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\822192540.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_2 = pd.read_sql_query(query_6_2, connection)


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [50]:
# текст запроса
query_6_3 = f'''
                SELECT count(*)
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']) --фильтр для поиска подходящих по названию вакансий
                OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
                AND (key_skills ILIKE ANY (array['%sql%', '%postgre%']))                                                                               --дополнительный фильтр для более узкого поиска
'''

In [51]:
# результат запроса
result_6_3 = pd.read_sql_query(query_6_3, connection)
print(result_6_3.iloc[0, 0])

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\569399340.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_3 = pd.read_sql_query(query_6_3, connection)


201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [52]:
# текст запроса
query_6_4 = f'''
                SELECT count(*)
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])
                OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
                AND (key_skills ILIKE '%python%')
'''

In [53]:
# результат запроса
result_6_4 = pd.read_sql_query(query_6_4, connection)
print(result_6_4.iloc[0, 0])

351


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\4193089791.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_4 = pd.read_sql_query(query_6_4, connection)


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [54]:
# текст запроса
query_6_5 = f'''
                SELECT avg((length(key_skills) - length(replace(key_skills, e'\t', '')))+1)     --ключевые навыки разделены табуляцией, считаем разность количества символов в строках с табуляцией и без.
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])
                OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
'''

In [55]:
# результат запроса
result_6_5 = pd.read_sql_query(query_6_5, connection)
print(result_6_5.iloc[0, 0].round(2))

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\700219386.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_5 = pd.read_sql_query(query_6_5, connection)


6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [56]:
# текст запроса
query_6_6 = f'''
                SELECT
                    experience,                                                         --опыт работы
                    avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to))  --средняя зарплата, так как в значениях есть NULL, используем функцию coalesce
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']) OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
                AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                GROUP BY experience
'''

In [57]:
# результат запроса
result_6_6 = pd.read_sql_query(query_6_6, connection)
print(result_6_6.round())

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\181823913.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_6 = pd.read_sql_query(query_6_6, connection)


           experience       avg
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0


***

In [58]:
# выводы по предметному анализу

In [59]:
# количество Data Scientist вакансий
query_ds = f'''
                SELECT count(*)                                                                                                                        --количество вакансий
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']) --фильтр для поиска подходящих по названию вакансий
                OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
'''

result_ds = pd.read_sql_query(query_ds, connection)
print(result_ds.iloc[0, 0])
# количество DS вакансий уровня Middle
query_midds = f'''
                SELECT count(*)                                                                                                                        --количество вакансий
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']) --фильтр для поиска подходящих по названию вакансий
                OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
                AND (name ILIKE '%middle%' OR experience = 'От 1 года до 3 лет')                                                                       --дополнительный фильтр для более узкого поиска
'''

result_midds = pd.read_sql_query(query_midds, connection)
print(result_midds.iloc[0, 0])

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1023634410.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_ds = pd.read_sql_query(query_ds, connection)


480


C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1023634410.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_midds = pd.read_sql_query(query_midds, connection)


243


Количество вакансий имеющих отношение к данным 1771, из них 480 вакансий по Data Science.
Число DS вакансий для уровня:
* Junior - 51
* Middle - 243
* Senior и прочее - 186

Около половины DS вакансий требуют знания SQL, а знания Python необходимы в 351 вакансии.

Средняя з.п. DS вакансий для уровня:
* Junior - 75 000
* Middle - 140 000
* Senior и прочее - 240 000

# Общий вывод по проекту

In [60]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### Дополнительное исследование
Давайте узнаем какими навыками должен обладать специалист Data Science.

Для этого сделаем запрос к нашей базе данных по ключевым навыкам (сделаем ТОП ключевых навыков).

In [61]:
query_topkeysk = f'''
            SELECT
                skill,                                                                                                                                  --ключевой навык
                count(*) cnt                                                                                                                            --количество
            FROM
                (SELECT regexp_split_to_table(replace(lower(key_skills), ' ', '_'), '\s+') skill                                                        --разделяем строку key_skill по табуляции и делаем таблицу из ключевых навыков
                 FROM public.vacancies
                 WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']) --фильтр по проффессии
                 OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
                ) AS s
            GROUP BY skill
            ORDER BY cnt DESC
'''

result_topkeysk = pd.read_sql_query(query_topkeysk, connection)
print(result_topkeysk.head(21))

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\4004135451.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_topkeysk = pd.read_sql_query(query_topkeysk, connection)


                           skill  cnt
0                         python  348
1                            sql  191
2               machine_learning  140
3                            git   66
4                   data_science   64
5      математическая_статистика   62
6                        pytorch   61
7                         pandas   55
8                  data_analysis   54
9                             ml   49
10               английский_язык   48
11                         linux   47
12                 анализ_данных   46
13                           nlp   42
14                    tensorflow   42
15                      big_data   42
16                        docker   41
17  математическое_моделирование   41
18         статистический_анализ   40
19                           c++   39
20                         numpy   38


В нашем запросе не учитываются небольшие различия в названиях навыков, но даже в таком виде становиться ясно какими навыками должен обладать DS специалист, а это:
* Python
* SQL
* Алгоритмы Machine Learning
* Git
* Мат. статистика
* Английский язык
* Pandas, NumPy, Pytorch, TensorFlow и т.д.

In [62]:
# Посмотрим сколько указано зарплат для DS
query_sal = f'''
                SELECT count(*)                                                                                                                        --количество вакансий
                FROM public.vacancies
                WHERE (name ILIKE ANY (array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']) --фильтр для поиска подходящих по названию вакансий
                OR (name NOT ILIKE '%HTML%' AND name like '%ML%'))
                AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)                                                                                 --фильтр по указанным зарплатам
'''

result_sal = pd.read_sql_query(query_sal, connection)
print(result_sal.iloc[0, 0])

C:\Users\admin\AppData\Local\Temp\ipykernel_14740\1238765384.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_sal = pd.read_sql_query(query_sal, connection)


62


In [63]:
# закрываем соединение
connection.close()

Как видим из 480 вакансий по Data Science, зарплата указанна только у 62.

---

### Общий вывод

Для нашего проекта: ***создание модели машинного обучения, которая будет рекомендовать вакансии клиентам агентства, претендующим на позицию Data Scientist***,

необходимо провести дальнейший анализ вакансий, выявить зависимости зарплаты от опыта работы, набора ключевых навыков, графика работы и т.д. и сначала, создать

модель машинного обучения для заполнения недостающих данных о заработной плате. Тогда наша база данных будет хорошо подходить для нашей основной цели - создание рекомендательной системы.